dotnet/sdk-phase-2.md

```diff
@@ -0,0 +1,484 @@
+
+* Every workflow must have the non-inheritable `[Workflow]` attribute
+  * Optional `Name` positional property can be given. Default is the unqualified type name with the "I" prefix removed
+    if it's an interface and the next letter is also capital
+    * So what does `IDWorkflow` become? Yes, `DWorkflow` :-( Weighed as an acceptably rare cost for obvious benefit
+  * Non-inheritable means just because the interface or base class defines `[Workflow]`, the registered one must too
+* Workflows can have a single public constructor with `[WorkflowInit]` that take the exact same parameters as the
+  `[WorkflowRun]` method
+  * This is not required and most workflows probably won't use it but it is nice for those wanting to build readonly
+    properties instead of waiting until run
+  * If this is not present, a public parameterless constructor must be present (default of course in C# when no
+    constructors present)
+  * Ignored if in base class
+* Every workflow must explicitly have one and only one public method with a non-inheritable `WorkflowRun` attribute
+  * Can be in base class, but must also be overridden in this workflow class w/ the attribute even if it just delegates
+    to super
+  * Method must return a `Task`
+* Workflow signals are public async methods with the non-inheritable `[WorkflowSignal]` attribute
+  * Can be in base class, but if overridden, must also have the attribute
+  * Method must return a `Task` and can accept arguments
+    * It is a bit annoying to force Task on people that may not be doing an async thing but:
+      * They may change it later so might as well require it (especially since the "Async" suffix is usually present)
+      * Causes too many overloads if we accept `Task` and `void`
+  * Optional `Name` positional property can be given. Default is the unqualified method name with "Async" trimmed off if
+    present.
+* Workflow queries are public methods with the non-inheritable `[WorkflowQuery]` attribute
+  * Can be in base class, but if overridden, must also have the attribute
+  * Method cannot return a `Task`, but can accept arguments
+  * Optional `Name` positional property can be given. Default is the unqualified method name.
+  * Would love to put this on properties, but they aren't referenceable as delegates
+* Users are encouraged to set a `static` `readonly` `Ref` property for use by clients
+  * `Refs.Create<T>()` is how it is obtained
+    * `Refs` is in `Temporalio` not `Temporalio.Workflows`
+      * TODO(cretz): This acceptable?
+    * TODO(cretz): Should we do validation at this time on the attributes?
+      * Problem is instance activities will want to use `Refs` too.
+      * If I put `Refs` in `Temporalio.Workflows` then at least I can know that all uses are for workflows
+        * But then it'd clash with a `Temporalio.Activities.Refs` or do we want something like either
+          `Refs.ForWorkflow<T>` or `WorkflowRefs.Create<T>`?
```

Comment:
I think `WorklowRefs.Create<T>` is best

Reply:
Better than `Refs.Create<T>()`?

Reply:
Was decided to separate this into `WorkflowRefs` and `ActivityRefs` that will also validate the workflow annotations.

Reply:
Fixed